src/coreclr/jit/morph.cpp

```diff
@@ -12651,79 +12662,199 @@ GenTree* Compiler::fgOptimizeEqualityComparisonWithConst(GenTreeOp* cmp)
 
-    if (op2Value == 1)
-    {
-        // Check for "expr >= 1".
-        if (oper == GT_GE)
+        genTreeOps oper     = cmp->OperGet();
+        int64_t    op2Value = op2->IntegralValue();
+
+        if (op2Value == 1)
         {
-            // Change to "expr != 0" for unsigned and "expr > 0" for signed.
-            oper = cmp->IsUnsigned() ? GT_NE : GT_GT;
+            // Check for "expr >= 1".
+            if (oper == GT_GE)
+            {
+                // Change to "expr != 0" for unsigned and "expr > 0" for signed.
+                oper = cmp->IsUnsigned() ? GT_NE : GT_GT;
+            }
+            // Check for "expr < 1".
+            else if (oper == GT_LT)
+            {
+                // Change to "expr == 0" for unsigned and "expr <= 0".
+                oper = cmp->IsUnsigned() ? GT_EQ : GT_LE;
+            }
         }
-        // Check for "expr < 1".
-        else if (oper == GT_LT)
+        // Check for "expr relop -1".
+        else if (!cmp->IsUnsigned() && (op2Value == -1))
         {
-            // Change to "expr == 0" for unsigned and "expr <= 0".
-            oper = cmp->IsUnsigned() ? GT_EQ : GT_LE;
+            // Check for "expr <= -1".
+            if (oper == GT_LE)
+            {
+                // Change to "expr < 0".
+                oper = GT_LT;
+            }
+            // Check for "expr > -1".
+            else if (oper == GT_GT)
+            {
+                // Change to "expr >= 0".
+                oper = GT_GE;
+            }
         }
-    }
-    // Check for "expr relop -1".
-    else if (!cmp->IsUnsigned() && (op2Value == -1))
-    {
-        // Check for "expr <= -1".
-        if (oper == GT_LE)
+        else if (cmp->IsUnsigned())
         {
-            // Change to "expr < 0".
-            oper = GT_LT;
+            if ((oper == GT_LE) || (oper == GT_GT))
+            {
+                if (op2Value == 0)
+                {
+                    // IL doesn't have a cne instruction so compilers use cgt.un instead. The JIT
+                    // recognizes certain patterns that involve GT_NE (e.g (x & 4) != 0) and fails
+                    // if GT_GT is used instead. Transform (x GT_GT.unsigned 0) into (x GT_NE 0)
+                    // and (x GT_LE.unsigned 0) into (x GT_EQ 0). The later case is rare, it sometimes
+                    // occurs as a result of branch inversion.
+                    oper = (oper == GT_LE) ? GT_EQ : GT_NE;
+                    cmp->gtFlags &= ~GTF_UNSIGNED;
+                }
+                // LE_UN/GT_UN(expr, int/long.MaxValue) => GE/LT(expr, 0).
+                else if (((op1->TypeIs(TYP_LONG) && (op2Value == INT64_MAX))) ||
+                         ((genActualType(op1) == TYP_INT) && (op2Value == INT32_MAX)))
+                {
+                    oper = (oper == GT_LE) ? GT_GE : GT_LT;
+                    cmp->gtFlags &= ~GTF_UNSIGNED;
+                }
+            }
         }
```

Comment:
Why are you changing all this?
I think you should rather leave this alone and add a new `fgFoldRelationalComparison` method containing your logic. That will be much easier to review.